rfcs/2021-09-21-9292-throttle-transform.md

```diff
@@ -0,0 +1,149 @@
+    key_field: Option<String>,
+    exclude: Option<AnyCondition>,
+}
+```
+
+TaskTransform:
+
+```rust
+impl TaskTransform for Throttle {
+    fn transform(
+        self: Box<Self>,
+        mut input_rx: Pin<Box<dyn Stream<Item = Event> + Send>>,
+    ) -> Pin<Box<dyn Stream<Item = Event> + Send>>
+    where
+        Self: 'static,
+    {
+        let lim = RateLimiter::keyed(Quota::per_second(self.events_per_second));
+
+        let mut flush_stream = tokio::time::interval(Duration::from_millis(1000));
+
+        Box::pin(
+            stream! {
+              loop {
+                let mut output = Vec::new();
+                let done = tokio::select! {
+                    _ = flush_stream.tick() => {
+                        false
+                    }
+                    maybe_event = input_rx.next() => {
+                        match maybe_event {
+                            None => true,
+                            Some(event) => {
+			    	if let Some(condition) = self.exclude_as_ref() {
+				    if condition.check(&event) {
+				        output.push(event);
+					false
+				    }
+				}
+
+				let value = self
+				    .key_field
+				    .as_ref()
+				    .and_then(|key_field| event.get(key_field))
+				    .map(|v| v.to_string_lossy());
+
+                                match lim.check_key_n(value, size) {
+                                    Ok(()) => {
+                                        output.push(event);
+                                        false
+                                    }
+                                    _ => {
+                                        // Dropping event
+                                        false
+                                    }
+                                }
+                            }
+                        }
+                    }
+                };
+                yield stream::iter(output.into_iter());
+                if done { break }
+              }
+            }
+            .flatten(),
+        )
+    }
+}
+```
+
+## Rationale
+
+- Controlling throughput provides better reliability for downstream services which can be critical in observability platforms
+- Setting granular quotas is invaluable as an administrator running Vector for a number of users to ensure fair use of the pipeline
+
+## Drawbacks
+
+- When rate limiting by `Bytes` in a transform it will be the unserialized form of the event, which will differ from what downstream sinks will actually receive
+
+## Prior Art
+
+- [Governor](https://docs.rs/governor/0.3.2/governor/index.html)
+- [Throttle - FluentBit](https://docs.fluentbit.io/manual/pipeline/filters/throttle)
+- [grouper::bucker - Tremor](https://github.com/tremor-rs/tremor-runtime/blob/main/tremor-pipeline/src/op/grouper/bucket.rs)
+
+## Alternatives
+
+- Extend `sample` transform to allow for a window configuration
+- Build rate limiting controls as a generic `sink` feature, rather than a separate `transform`
```

Comment:
This is interesting, as it would potentially resolve the "drawback" you listed above, if we think it's a drawback worth solving.

It would be good to add a bit more context to both alternatives; were they discussed before, what was the reason to list them as alternatives as opposed to making one of the two the primary choice, and having the `throttle` transform be one of the alternatives?

Reply:
😄 @binarylogic brought up the sink feature today, I had already gotten a working spike and most of the RFC done and hadn't considered a composable feature for sinks. It would tie in somewhat with ARC which could be difficult to optimize?

I'd also note that today we're already not exact with payload size, the new `datalog_logs` sink limits the payload size to an arbitrarily smaller size than the API can actually receive because of the serialized size problem (IIRC).

Reply:
I would suggest, even if we do the sink-based throttling too, that we could support the transform as well.